packages/editor/src/lib/utils/shapeNavigation.ts

```diff
@@ -0,0 +1,199 @@
+import { TLShape, TLShapeId } from '@tldraw/tlschema'
+import { Editor } from '../editor/Editor'
+import { Vec } from '../primitives/Vec'
+
+const SHALLOW_ANGLE = 20
+const ROW_THRESHOLD = 100
+
+/** @public */
+export type AdjacentDirection = 'next' | 'prev' | 'left' | 'right' | 'up' | 'down'
+
+/** @public */
+export function selectAdjacentShape(editor: Editor, direction: AdjacentDirection) {
+	const currentShapeId = editor.getSelectedShapeIds()[0]
+
+	let shapes: TLShape[] = []
+	let adjacentShapeId: TLShapeId
+	if (direction === 'next' || direction === 'prev') {
+		shapes = getShapesInReadingOrder(editor)
+		const shapeIds = shapes.map((shape) => shape.id)
+
+		const currentIndex = shapeIds.indexOf(currentShapeId)
+		const adjacentIndex =
+			(currentIndex + (['next', 'right', 'down'].includes(direction) ? 1 : -1) + shapeIds.length) %
+			shapeIds.length
+		adjacentShapeId = shapeIds[adjacentIndex]
+	} else {
+		if (!currentShapeId) return
+		adjacentShapeId = getNearestAdjacentShape(editor, currentShapeId, direction)
+	}
+
+	const shape = editor.getShape(adjacentShapeId)
+	if (!shape) return
+
+	editor.setSelectedShapes([shape.id])
+	editor.zoomToShapeIfOffscreen(256)
+}
+
+/**
+ * Generates a reading order for shapes based on rows grouping.
+ * Tries to keep a natural reading order (left-to-right, top-to-bottom).
+ *
+ * @public
+ */
+export function getShapesInReadingOrder(editor: Editor): TLShape[] {
```

Comment:
Can we rename this to `getCurrentPageShapesInReadingOrder`?

Also, is it worth sticking this in a `computed` so it's cached between invocations? I can imagine this getting expensive on big documents. 

If you keep this as a free function, you can do this with:
```
const shapesInReadingOrderStore = new WeakCache<Editor, Computed<TLShape[]>>()
export function getShapesInReadingOrder(editor: Editor) {
  return shapesInReadingOrderStore.get(editor, () => computed('shapes in reading order', () => {
    // existing impl goes here
  })).get()
}
```

or if you make it a method u can just stick `@computed` on it